RecoTauTag/RecoTau/plugins/DeepTauId.cc

```diff
@@ -587,33 +607,25 @@ namespace {
   };
 
   namespace candFunc {
-    auto getTauDz(const reco::PFCandidate& cand, float default_value) {
-      return cand.bestTrack() != nullptr ? cand.bestTrack()->dz() : default_value;
-    }
-    auto getTauDz(const pat::PackedCandidate& cand, float default_value) { return cand.dz(); }
-    auto getTauDzError(const reco::PFCandidate& cand, float default_value) {
-      return cand.bestTrack() != nullptr ? cand.dzError() : default_value;
-    }
-    auto getTauDzError(const pat::PackedCandidate& cand, float default_value) {
-      return cand.hasTrackDetails() ? cand.dzError() : default_value;
-    }
+    auto getTauDz(const reco::PFCandidate& cand) { return cand.bestTrack()->dz(); }
+    auto getTauDz(const pat::PackedCandidate& cand) { return cand.dz(); }
     auto getTauDZSigValid(const reco::PFCandidate& cand) {
       return cand.bestTrack() != nullptr && std::isnormal(cand.bestTrack()->dz()) && std::isnormal(cand.dzError()) &&
              cand.dzError() > 0;
     }
     auto getTauDZSigValid(const pat::PackedCandidate& cand) {
       return cand.hasTrackDetails() && std::isnormal(cand.dz()) && std::isnormal(cand.dzError()) && cand.dzError() > 0;
     }
-    auto getTauDxy(const reco::PFCandidate& cand, float default_value) {
-      return cand.bestTrack() != nullptr ? cand.bestTrack()->dxy() : default_value;
-    }
-    auto getTauDxy(const pat::PackedCandidate& cand, float default_value) { return cand.dxy(); }
+    auto getTauDxy(const reco::PFCandidate& cand) { return cand.bestTrack()->dxy(); }
+    auto getTauDxy(const pat::PackedCandidate& cand) { return cand.dxy(); }
     auto getPvAssocationQuality(const reco::PFCandidate& cand) { return 0.7013f; }
     auto getPvAssocationQuality(const pat::PackedCandidate& cand) { return cand.pvAssociationQuality(); }
-    auto getPuppiWeight(const reco::PFCandidate& cand) { return 0.9907f; }
-    auto getPuppiWeight(const pat::PackedCandidate& cand) { return cand.puppiWeight(); }
-    auto getPuppiWeightNoLep(const reco::PFCandidate& cand) { return 0.8858f; }
-    auto getPuppiWeightNoLep(const pat::PackedCandidate& cand) { return cand.puppiWeightNoLep(); }
+    auto getPuppiWeight(const reco::PFCandidate& cand, const float aod_value) { return aod_value; }
```

Comment:
BTW, why is this setup not allowing to use actual puppi weights? We are running PUPPI in reco and weights are available for all PFCands by the time the tau reco/ID can run.

Reply:
If this is the case, it was implemented like this because I overlooked it. I was under the impression that this was not possible, but if it is, it should probably be changed. @kandrosov To double-check, should these values be kept because it could otherwise give strange results or should it be fine to change this if it is possible?

Reply:
PUPPI weights are not available at the HLT level, as far as I know. For the moment, the only use-case to run DeepTau with AOD inputs is HLT. We can pass `is_online_` boolean to this (and other) method to make it more explicit.